Content.Server/CartridgeLoader/Cartridges/HealthAnalyzerCartridgeSystem.cs

```diff
@@ -0,0 +1,75 @@
+using Content.Shared.Damage;
+using Content.Shared.DoAfter;
+using Content.Shared.MedicalScanner;
+
+namespace Content.Server.CartridgeLoader.Cartridges;
+
+/// <summary>
+/// Handles install and removal of Health Analyzer Cartridge into a PDA
+/// </summary>
+public sealed class HealthAnalyzerCartridgeSystem : EntitySystem
+{
+    [Dependency] private readonly CartridgeLoaderSystem? _cartridgeLoaderSystem = default!;
+    [Dependency] private readonly SharedAudioSystem _audio = default!;
+    [Dependency] private readonly SharedDoAfterSystem _doAfterSystem = default!;
+
+    public override void Initialize()
+    {
+        base.Initialize();
+        SubscribeLocalEvent<HealthAnalyzerCartridgeComponent, CartridgeUiReadyEvent>(OnUiReady);
+        SubscribeLocalEvent<HealthAnalyzerCartridgeComponent, CartridgeAfterInteractEvent>(AfterInteract);
+        SubscribeLocalEvent<HealthAnalyzerCartridgeComponent, HealthAnalyzerDoAfterEvent>(OnDoAfter);
+    }
+
+    private void AfterInteract(EntityUid uid, HealthAnalyzerCartridgeComponent component, CartridgeAfterInteractEvent args)
+    {
+        if (args.InteractEvent.Handled || !args.InteractEvent.CanReach || !args.InteractEvent.Target.HasValue)
+            return;
+
+        _audio.PlayPvs(component.ScanningBeginSound, uid);
+
+        _doAfterSystem.TryStartDoAfter(new DoAfterArgs(args.Loader, component.ScanDelay, new HealthAnalyzerDoAfterEvent(), uid, target: args.InteractEvent.Target, used: args.Loader)
+        {
+            BreakOnTargetMove = true,
+            BreakOnUserMove = true,
+            NeedHand = false
+        });
+    }
+
+    private void OnDoAfter(EntityUid uid, HealthAnalyzerCartridgeComponent component, DoAfterEvent args)
+    {
+        if (args.Handled || args.Cancelled || args.Args.Target == null || args.Args.Used == null)
+            return;
+
+        _audio.PlayPvs(component.ScanningEndSound, uid);
+
+        UpdateScannedUser(uid, args.Args.User, args.Args.Target, component);
+        args.Handled = true;
+    }
+
+    private void OnUiReady(EntityUid uid, HealthAnalyzerCartridgeComponent component, CartridgeUiReadyEvent args)
+    {
+        UpdateScannedUser(uid, args.Loader, null, component);
+    }
+
+    private void UpdateScannedUser(EntityUid uid, EntityUid loaderUid, EntityUid? target, HealthAnalyzerCartridgeComponent? component)
+    {
+        if (!Resolve(uid, ref component))
+            return;
+
+        if (!HasComp<DamageableComponent>(target))
+            return;
+
+        TryComp<TemperatureComponent>(target, out var temp);
+        TryComp<BloodstreamComponent>(target, out var bloodstream);
+
+        var scannedUserMessage = new HealthAnalyzerScannedUserMessage(target,
+            temp?.CurrentTemperature ?? float.NaN,
+            bloodstream != null ? bloodstream.BloodSolution.FillFraction : float.NaN);
+        _cartridgeLoaderSystem?.UpdateCartridgeUiState(loaderUid, new HealthAnalyzerUiState(scannedUserMessage));
+    }
+}
```

Comment:
Remove `UpdateScannedUser`, add the `HealthAnalyzerSystem` as a dependency and call `_healthAnalzerSystem.UpdateScannedUser(...)` instead of just copying the method from the system into here

Reply:
I'm a little confused how I would have the `HealthAnalyzerSystem` know whether to call the `_cartridgeLoaderSystem?.UpdateCartridgeUiState` or `_uiSystem.SendUiMessage` if I did this. Is there a way I can have the `CartridgeLoaderSystem` pick up on the `UiSystem`'s `SendUiMessage`?

Reply:
Change the health analyzer system to use ui states instead of messages